tvb/recon/flow/surfaces_to_structural_datasets.py

```diff
@@ -20,6 +21,131 @@ class Hemisphere(enum.Enum):
+                (verts[ang[2], :] - verts[ang[0], :]) /
+                np.sqrt(np.sum((verts[ang[2], :] - verts[ang[0], :]) ** 2, axis=0))))
+
+    return angles
+
+
+def compute_vertex_normals(number_of_vertices, vertex_triangles, triangles,
+                           triangle_angles, triangle_normals, vertices):
+    """
+    Estimates vertex normals, based on triangle normals weighted by the
+    angle they subtend at each vertex...
+    """
+    vert_norms = np.zeros((number_of_vertices, 3))
+    bad_normal_count = 0
+    for k in range(number_of_vertices):
+        try:
+            tri_list = list(vertex_triangles[k])
+            angle_mask = triangles[tri_list, :] == k
+            angles = triangle_angles[tri_list, :]
+            angles = angles[angle_mask][:, np.newaxis]
+            angle_scaling = angles / np.sum(angles, axis=0)
+            vert_norms[k, :] = np.mean(angle_scaling * triangle_normals[tri_list, :], axis=0)
+            # Scale by angle subtended.
+            vert_norms[k, :] = vert_norms[k, :] / np.sqrt(np.sum(vert_norms[k, :] ** 2, axis=0))
+            # Normalise to unit vectors.
+        except (ValueError, FloatingPointError):
+            # If normals are bad, default to position vector
+            # A nicer solution would be to detect degenerate triangles and ignore their
+            # contribution to the vertex normal
+            vert_norms[k, :] = vertices[k] / np.sqrt(vertices[k].dot(vertices[k]))
+            bad_normal_count += 1
+    if bad_normal_count:
+        print(" %d vertices have bad normals" % bad_normal_count)
+    return vert_norms
+
+
+def compute_region_orientations(regions, vertex_normals, region_mapping):
+    """Compute the orientation of given regions from vertex_normals and region mapping"""
+
+    average_orientation = np.zeros(regions.size, 3)
+    # Average orientation of the region
+    for i, k in enumerate(regions):
+        orient = vertex_normals[region_mapping == k, :]
+        if orient:
+            avg_orient = np.mean(orient, axis=0)
+            average_orientation[i, :] = avg_orient / np.sqrt(np.sum(avg_orient ** 2))
+
+    return average_orientation
+
+
+def compute_triangle_areas(vertices, triangles):
+    """Calculates the area of triangles making up a surface."""
+    tri_u = vertices[triangles[:, 1], :] - vertices[triangles[:, 0], :]
+    tri_v = vertices[triangles[:, 2], :] - vertices[triangles[:, 0], :]
+    tri_norm = np.cross(tri_u, tri_v)
+    triangle_areas = np.sqrt(np.sum(tri_norm ** 2, axis=1)) / 2.0
+    triangle_areas = triangle_areas[:, np.newaxis]
+    return triangle_areas
+
+
+def compute_region_areas(regions, triangle_areas, vertex_triangles, region_mapping):
+    """Compute the areas of given regions"""
+
+    region_surface_area = np.zeros(regions.size, 1)
+    avt = np.array(vertex_triangles)
+    # NOTE: Slightly overestimates as it counts overlapping border triangles,
```

Comment:
I wonder how FS handles that?  I guess their triangles are so small, they can exclude borders and be fine.